discord/utils.py

```diff
@@ -567,64 +585,167 @@ def get(iterable: Iterable[T], **attrs: Any) -> T | None:
     return None
 
 
-async def get_or_fetch(obj, attr: str, id: int, *, default: Any = MISSING) -> Any:
-    """|coro|
+_FETCHABLE = TypeVar(
+    "_FETCHABLE",
+    bound="VoiceChannel | TextChannel | ForumChannel | StageChannel | CategoryChannel | Thread | Member | User | Guild | Role | GuildEmoji | AppEmoji",
+)
 
-    Attempts to get an attribute from the object in cache. If it fails, it will attempt to fetch it.
-    If the fetch also fails, an error will be raised.
+
+# TODO: In version 3.0, remove the 'attr' and 'id' arguments.
+#       Also, eliminate the default 'MISSING' value for both 'object_type' and 'object_id'.
+async def get_or_fetch(
+    obj: Guild | Client,
+    object_type: type[_FETCHABLE] = MISSING,
+    object_id: int | None = MISSING,
+    default: Any = MISSING,
+    attr: str = MISSING,
+    id: int = MISSING,
+) -> _FETCHABLE | None:
```

Comment:
It would be nice if the default parameter’s type was also included here, as that’s what’s returned when specified. Otherwise, it raises an error. 

And maybe overload on `object_id` since that's when it always returns `None` (from what I can see). Otherwise it returns `<default>` or raises whatever the specific fetch method raised.